File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SaintLucia.m
//
//  UIImage+SaintLucia.m
//  CountriesFlags
//

#import "Countries/UIImage+SaintLucia.h"

@implementation UIImage (SaintLucia)

+ (UIImage * _Nullable)flagSaintLuciaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:102.0f/255.0f
                                         green:204.0f/255.0f
                                          blue:1.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorBlack = [UIColor blackColor];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:209.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(size.width / 72.0f * 24.0f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 36.0f * 3.16760303)];
    [path addLineToPoint:CGPointMake(size.width / 72.0f * 48.0f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path closePath];
    [path fill];

    [colorBlack setFill];
    path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(size.width / 72.0f * 25.6f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 36.0f * 7.16760303)];
    [path addLineToPoint:CGPointMake(size.width / 72.0f * 46.4f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path closePath];
    [path fill];

    [colorYellow setFill];
    path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(size.width / 72.0f * 24.0f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(size.width / 72.0f * 48.0f, size.height / 36.0f * 14.83239697 * 2.0f + size.height / 36.0f * 3.16760303)];
    [path closePath];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SaintLucia.swift
//
//  UIImage+SaintLucia.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Saint_Lucia "wikipedia"
     
     Generates the flag of Saint Lucia.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Saint Lucia.
     */
    internal static func flagSaintLucia(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 102.0 / 255.0,
                                green: 204.0 / 255.0,
                                blue: 1.0,
                                alpha: 1.0)
        let colorWhite = UIColor.white
        let colorBlack = UIColor.black
        let colorYellow = UIColor(red: 1.0,
                                  green: 209.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        let xUnit = size.width / 72.0
        let yUnit = size.height / 36.0
        let path = UIBezierPath()
        path.move(to: CGPoint(x: xUnit * 24.0, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: yUnit * 3.16760303))
        path.addLine(to: CGPoint(x: xUnit * 48.0, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.close()
        path.fill()

        colorBlack.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: xUnit * 25.6, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: yUnit * 7.16760303))
        path.addLine(to: CGPoint(x: xUnit * 46.4, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.close()
        path.fill()

        colorYellow.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: xUnit * 24.0, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: xUnit * 48.0, y: yUnit * 14.83239697 * 2.0 + yUnit * 3.16760303))
        path.close()
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
